Bearded.Utilities/Noise/NoiseMap.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+
+namespace Bearded.Utilities.Noise
+{
+    public static class NoiseMap
+    {
+        public static INoiseMap FromArray(double[,] array, IInterpolationMethod2 interpolation) =>
+            new ArrayNoiseMap(array, interpolation);
+
+        private sealed class ArrayNoiseMap : INoiseMap
+        {
+            private readonly double[,] array;
+            private readonly IInterpolationMethod2 interpolation;
+            private readonly int width;
+            private readonly int height;
+
+            public ArrayNoiseMap(double[,] array, IInterpolationMethod2 interpolation)
+            {
+                this.array = array;
+                this.interpolation = interpolation;
+                width = array.GetLength(0);
+                height = array.GetLength(1);
+            }
+
+            public double ValueAt(double x, double y)
+            {
+                if (x < 0 || x >= 1)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(x));
+                }
+                if (y < 0 || y >= 1)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(y));
+                }
+
+                // First we transform x and y from [0, 1) x [0, 1) to [0, width) x [0, height) for easier math.
+                x *= width;
+                y *= height;
+
+                // The noise array keeps track of the values at the center of each "pixel", so for the interpolation, we
+                // need to move the array coordinates by (0.5, 0.5), which is the same as moving the x, y by
+                // (-0.5, -0.5).
+                x -= 0.5;
+                y -= 0.5;
+
+                var xBelow = MoreMath.FloorToInt(x);
+                var xAbove = MoreMath.CeilToInt(x);
+                var yBelow = MoreMath.FloorToInt(y);
+                var yAbove = MoreMath.CeilToInt(y);
+                var u = x - xBelow;
+                var v = y - yBelow;
+
+                xBelow = (xBelow + width) % width;
```

Comment:
So this actually wraps around? Is that really intended? I definitely didn't expect it. I would have thought it would map 0 to the first pixel and 1 to the last, instead both map to the average of first and last?

I'm not sure if wrapping is good default behaviour. If anything maybe we have to inject it, if we want to support it?

Reply:
Imagine a simple noise texture as actual pixels. The array here represents the pixels of that texture. The value you pass in is actually the continuous coordinate on that texture, and there's a good chance that coordinate doesn't exactly align with the centre of a pixel. This is where the interpolation method that gets passed in comes in. If you do bilinear interpolation, you need something to interpolate with. We could interpolate with 0, but this causes artefacts along the outside, which I think is undesirable.

Note that if you use nearest neighbour as interpolation method, this will all still add up to returning the value of pixel [0, 0] in coordinates (0, 0).

> I'm not sure if wrapping is good default behaviour. If anything maybe we have to inject it, if we want to support it?

I agree the wrapping behaviour should be injected, but given the already existent complexity of this PR, has been deferred to future PRs. I think it makes sense to provide a default value though, and I chose repeat in this case (which is in line with the [OpenGL texture default](https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glTexParameter.xml))